### PR TITLE
[CFDP-1119] [CFDP-1190] Cache Flush Enhancements

### PR DESCRIPTION
**Pull request by asleepingpanda**
## DESCRIPTION
Introduces a CacheManager datasource for these more complex interactions.

Recursively goes through an entity's attributes and flushes the cache for both the entity and attributes

### 1. What does this PR do, or why is it needed?
Because of how aggressive we are with caching data, we made a deal with out content team that we would give them the ability to manually clear the cache for specific items.

There were 2 main problems we've been facing recently:
* Cache clearing has only been working in stage, not production
* If a Content Item has a Schedule attached to it, the Content Item would clear its cache, but the associated schedule would not

This PR introduces a `Cache Manager` that handles recursively clearing out a Rock Entity's cache as well as the cache of the attribute values associated with it.

The reason that a `Cache Manager` was introduced as a new data source was simple for code organization.

This area of the API does not need to be particularly performant as it's to be treated as a background process, so there are plenty of plain 'ole for loops and switch cases all over the place. This was done with the goal of making the code a bit easier to read.

To Test:
Make sure you're connected to a Redis instance and watch the logs while running this mutation.
```
mutation flush($entityId: Int!, $entityTypeId: Int!, $apollosAccessKey: String!) {
  flushRock(
    entityId: $entityId, 
    entityTypeId: $entityTypeId, 
    key: $apollosAccessKey)
}
```

### CODE REVIEW

File: src/data/redis-cache-manager/data-source.js [hunk 1]
```diff
@@ -0,0 +1,213 @@
...
+      switch (entityTypeId) {
+        case CONTENT_CHANNEL:
+          log_cache('Content Channel', entityId);
+
+          await this.delete({
+            key: this.KEY_TEMPLATES.contentChannelItemIds`${entityId}`,
+          });
+          break;
+        case CONTENT_CHANNEL_ITEM:
+          log_cache('Content Channel Item', entityId);
```
Comment (rkdavidson) [inline src/data/redis-cache-manager/data-source.js:150]: Nitpick, but the `snake_case` feels awkward. Harmless, though.

File: src/data/redis-cache/resolver.js [hunk 2]
```diff
@@ -6,9 +6,15 @@ const { APOLLOS_SECRET } = ROCK;
 
 export default {
   Mutation: {
-    flushRock: (root, { key, ...args }, { dataSources: { Cache } }) => {
+    flushRock: async (root, { key, ...args }, { dataSources: { CacheManager } }) => {
       if (key === APOLLOS_SECRET) {
-        return Cache.flushFor('ROCK', args);
+        console.log('\n\x1b[35m[redis cache] starting recursive clear');
+
+        await CacheManager.recursivelyClear(args);
```
Comment (rkdavidson) [inline src/data/redis-cache/resolver.js:13]: Should we wrap this in a `try/catch` so it's extra-defensive?
The recursive nature makes me a bit nervous, and could hypothetically cripple the server if something funky and hard to reason about occurs.

It feels better to `try` the recursive clear, and if _anything_ goes wrong, simply log it and hope nothing broke?

**Review by rkdavidson — COMMENTED** [review]
I didn't test locally or manually check redis cache, etc, but conceptually the code seems to make sense.
Few comments.

File: src/data/redis-cache/resolver.js (hunk 2 above)
Comment (asleepingpanda) [reply src/data/redis-cache/resolver.js:13]: Yeah that's not a bad idea! I'll do that real quick

File: src/data/redis-cache-manager/data-source.js (hunk 1 above)
Comment (asleepingpanda) [reply src/data/redis-cache-manager/data-source.js:150]: Not a bad callout! Easy fix

**Comment by asleepingpanda** [comment]
@rkdavidson threw some resolution to those feedback items real quick

**Comment by rkdavidson** [comment]
Changes look good to me! Again, haven't tested locally, so if you're confident then I'm confident.
After we deploy, we can give it a shot with a group and content item (or something) to smoke test.

**Pull request merged**